### PR TITLE
Feat/boat manager

### LINKED ISSUE #56: Fix Remaining Memory Propogation issues

### PR DESCRIPTION
**Pull request by Eein**
closes #56

- Adds Boat Manager and updates NavHelper
- Fixes the annoying crash bug and updates PlayerPartyManager 

I'd prefer merging this in sooner than later just to avoid a weird rebase later. 

## Notes

f32 was used directly for better CPU cache pipelining.

When doing exploration for this package, I used the `quaternion-core` package but it resulted in some uncomfortable code changes and multiple Vector3 implementations.

It also required some advanced implementation for euler_angles that I did not understand and could not find an exact implementation compared to what we were doing in the python tas.

https://docs.rs/quaternion-core/latest/quaternion_core/fn.to_euler_angles.html

I implemented the Python Quaternion object in its stead.

### CODE REVIEW

**Comment by Eein** [comment]
Working on fixing a crash that happens when values dont exist.

Edit: Fixed !

File: src/memory/player_party_manager.rs [hunk 1]
```diff
@@ -24,7 +24,7 @@ impl Default for MemoryManager<PlayerPartyManagerData> {
 pub struct PlayerPartyManagerData {
     pub position: Vector3<f32>,
     pub gameobject_position: Vector3<f32>,
-    pub leader_offset: Option<u32>,
+    leader_offset: Option<u32>,
```
Comment (Eein) [inline src/memory/player_party_manager.rs:25]: this field does not need to be exposed

File: src/util/quaternion.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
+use libm::Libm;
+use std::f32::consts::PI;
+use vec3_rs::Vector3;
+
+#[derive(Default, Debug)]
```
Comment (orkaboy) [inline src/util/quaternion.rs:5]: Possible thing here would be to implement `Quaternion<T>` to get `f64` also.... but idk if that's useful to us. It's not like anything other than `f32/f64` would actually work. Dunno if there's a "`Real`" trait in the `std` library to bound it.

Feel free to ignore.

File: src/util/math.rs
Comment (orkaboy) [inline src/util/math.rs:1]: Is this file dead code? I think you removed `quaternion_core` unless that's in `std`. Also not included in `mod.rs`.

File: src/gui/helpers/nav_helper.rs [hunk 3]
```diff
@@ -111,7 +113,7 @@ impl GuiHelper for NavHelper {
         egui::CollapsingHeader::new("Boat Coordinates NOT IMPLEMENTED")
```
Comment (orkaboy) [inline src/gui/helpers/nav_helper.rs]: Presumably actually implemented now, if it works. Need to play the game for a few hours more to test it :P

File: src/memory/boat_manager.rs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+use memory::memory_manager::il2cpp::UnityMemoryManager;
+use memory::process::MemoryError;
+
+impl Default for MemoryManager<BoatManagerData> {
+    fn default() -> Self {
+        let manager = Self {
+            name: "BoatManager".to_string(),
+            data: BoatManagerData::default(),
+            manager: UnityMemoryManager {
```
Comment (orkaboy) [inline src/memory/boat_manager.rs]: Could presumably just be `manager: UnityMemoryManager::default(),`

File: src/util/math.rs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+}
+
+// Quaternion - Get Yaw Rotation
+pub fn to_yaw(q: Quaternion<f32>) -> f32 {
+    let angles = to_euler_angles(RotationType::Intrinsic, RotationSequence::XYZ, q);
+    if angles[0] > 0.0 {
+        return angles[1] - PI / 2.0;
+    }
+    PI - (angles[1] + PI / 2.0)
+}
```
Comment (Eein) [inline src/util/math.rs]: ignore this file - needs to be deleted

File: src/memory/boat_manager.rs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            let z = memory_context.read_pointer::<f32>(boat_rotation_ptr + 0x8)?;
+            let w = memory_context.read_pointer::<f32>(boat_rotation_ptr + 0xC)?;
+
+            self.rotation = Quaternion { x, y, z, w };
+        }
+
+        Ok(())
+    }
+
+    pub fn update_speed(&mut self, memory_context: &MemoryContext) -> Result<(), MemoryError> {
```
Comment (orkaboy) [inline src/memory/boat_manager.rs:96]: I'll take your word for it that the field offsets are named correctly :)

**Review by orkaboy — APPROVED** [review]

File: src/memory/boat_manager.rs (hunk 6 above)
Comment (Eein) [reply src/memory/boat_manager.rs:96]: i tested them!
This is one of the only places in the game with these weird backing field field names.

**Review by orkaboy — APPROVED** [review]

**Pull request merged**